### PR TITLE
Add detailed errors for `tool.uv.sources` deserialization failures

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Closes https://github.com/astral-sh/uv/issues/7817.

### CODE REVIEW

File: crates/uv-distribution/src/metadata/requires_dist.rs [hunk 1]
```diff
@@ -371,8 +365,10 @@ mod test {
           |
         8 | tqdm = { url = "§invalid#+#*Ä" }
           |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
-        data did not match any variant of untagged enum Source
+        invalid value: string "§invalid#+#*Ä", expected relative URL without a base
+
 
+        in `url`
```
Comment (charliermarsh) [inline crates/uv-distribution/src/metadata/requires_dist.rs:371]: This part I don't quite understand. It looks like https://github.com/dtolnay/serde-untagged/issues/2.

Comment (charliermarsh) [reply crates/uv-distribution/src/metadata/requires_dist.rs:375]: This is really bothering me...

**Review by konstin — APPROVED** [review]

File: crates/uv-distribution/src/metadata/requires_dist.rs (hunk 1 above)
Comment (BurntSushi) [reply crates/uv-distribution/src/metadata/requires_dist.rs:375]: Yeah annoying, but still an improvement. :)

A solution here isn't obvious to me. (Other than hand-rolling a `Deserialize` impl.)

**Review by BurntSushi — APPROVED** [review]
Nice!

**Pull request merged**